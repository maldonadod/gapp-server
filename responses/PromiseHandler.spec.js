const {
  handlerPromiseFactory
  ,handlerPromise
  ,handlerPromisePagination
} = require('./PromiseHandler')

const {
  parseResponse
} = require('../src/Pagination/business')

const {
  success
  ,error
} = require('./index')

const req = {
  body: {
    name: 'Homer Simpson'
  }
}

const reqFactory = data => Object.assign({}, data)

// FACTORIES
const getPromiseFactory = () => jest.fn(req => Promise.resolve(req.body))
const getPromiseRejectedFactory = () => jest.fn(req => Promise.reject(req.body))
const resFactory = (fn) => ({
  send: jest.fn(fn)
})

describe('handlerPromise', () => {
  test('getPromise should be called in order to get a promise', () => {
    const getPromise = getPromiseFactory()
    const handler = handlerPromise(getPromise)
    const res = resFactory()
    handler(req, res)
    .then(() => expect(getPromise.mock.calls.length).toEqual(1))
  })
  test('res should be called once on success and response model should match', () => {
    const handler = handlerPromise(getPromiseFactory())
    const res = resFactory(data => expect(data).toMatchSnapshot())
    handler(req, res)
    .then(() => expect(res.send.mock.calls.length).toEqual(1))
  })
  test('res should be called once on error and response model should match', () => {
    const handler = handlerPromise(getPromiseRejectedFactory())
    const res = resFactory(data => expect(data).toMatchSnapshot())
    handler(req, res)
    .then(() => expect(res.send.mock.calls.length).toEqual(1))
  })
})

describe('handlerPromisePagination', () => {
  test('the response match with Pagination.parseResponse', () => {
    const handler = handlerPromisePagination(getPromiseFactory())
    const res = resFactory(data => expect(data).toMatchSnapshot())
    const body = {
      docs: []
      ,total: 0
      ,offset: 0
      ,limit: 10
    }
    const req = reqFactory({
      body
    })
    handler(req, res)
  })
})
