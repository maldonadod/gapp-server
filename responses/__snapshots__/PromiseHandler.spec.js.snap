// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlerPromise res should be called once on error and response model should match 1`] = `
Object {
  "message": Object {
    "name": "Homer Simpson",
  },
  "status": "ERR",
}
`;

exports[`handlerPromise res should be called once on success and response model should match 1`] = `
Object {
  "data": Object {
    "name": "Homer Simpson",
  },
  "status": "OK",
}
`;

exports[`handlerPromisePagination the response match with Pagination.parseResponse 1`] = `
Object {
  "data": Object {
    "paging": Object {
      "limit": 10,
      "offset": 0,
      "total": 0,
    },
    "results": Array [],
  },
  "status": "OK",
}
`;
